gradio/components/image.py

```diff
@@ -160,76 +147,26 @@ def __init__(
 
-    def _format_image(
-        self, im: _Image.Image | None
-    ) -> np.ndarray | _Image.Image | str | None:
-        """Helper method to format an image based on self.type"""
-        if im is None:
-            return im
-        fmt = im.format
-        if self.type == "pil":
-            return im
-        elif self.type == "numpy":
-            return np.array(im)
-        elif self.type == "filepath":
-            path = processing_utils.save_pil_to_cache(
-                im, cache_dir=self.GRADIO_CACHE, format=fmt or "png"  # type: ignore
-            )
-            return path
-        else:
-            raise ValueError(
-                "Unknown type: "
-                + str(self.type)
-                + ". Please choose from: 'numpy', 'pil', 'filepath'."
-            )
-
-    def preprocess(
-        self, x: str | dict[str, str]
-    ) -> np.ndarray | _Image.Image | str | dict | None:
+    def preprocess(self, x: FileData | None) -> np.ndarray | _Image.Image | str | None:
         """
         Parameters:
-            x: base64 url data, or (if tool == "sketch") a dict of image and mask base64 url data
+            x: FileData containing an image path pointing to the user's image
         Returns:
             image in requested format, or (if tool == "sketch") a dict of image and mask in requested format
         """
         if x is None:
             return x
 
-        mask = ""
-        if self.tool == "sketch" and self.source in ["upload", "webcam"]:
-            assert isinstance(x, dict)
-            x, mask = x["image"], x["mask"]
+        print(x)
 
-        if isinstance(x, str):
-            im = processing_utils.decode_base64_to_image(x)
-        else:
-            im = _Image.open(x["name"])
+        im = _Image.open(x["path"])
         with warnings.catch_warnings():
             warnings.simplefilter("ignore")
             im = im.convert(self.image_mode)
-        if self.shape is not None:
-            im = processing_utils.resize_and_crop(im, self.shape)
-        if self.invert_colors:
-            im = PIL.ImageOps.invert(im)
-        if (
-            self.source == "webcam"
-            and self.mirror_webcam is True
-            and self.tool != "color-sketch"
-        ):
-            im = PIL.ImageOps.mirror(im)
-
-        if self.tool == "sketch" and self.source in ["upload", "webcam"]:
-            mask_im = processing_utils.decode_base64_to_image(mask)
 
-            if mask_im.mode == "RGBA":  # whiten any opaque pixels in the mask
-                alpha_data = mask_im.getchannel("A").convert("L")
-                mask_im = _Image.merge("RGB", [alpha_data, alpha_data, alpha_data])
-            return {
-                "image": self._format_image(im),
-                "mask": self._format_image(mask_im),
-            }
-
-        return self._format_image(im)
+        return imagetools.format_image(
+            im, cast(Literal["numpy", "pil", "filepath"], self.type), self.GRADIO_CACHE
+        )
```

Comment:
I move this into a utils file because i'll need it for `ImageEditor` too and it cleans up the component a bit. Had to cast cos the typing sucks

Reply:
Nice